Multiply big integers via ARITH

MCJOHN974 opened the pull request:
Currently multiplication works only for small immediates via ``*``. Here I added support of multiplication for all integers via ``ARITH`` instruction and test for multiplication of big numbers

MCJOHN974 commented:
@nagisa, can you pls check this PR and questions above?

nagisa commented on cranelift/filetests/Cargo.toml:
```diff
@@ -36,3 +36,7 @@ cranelift-wasm.workspace = true
 expect-test = "1.4.1"
+
+[dev-dependencies]
+env_logger = "0.10"
```
Use `.workspace` dependencies instead. While at it, if you could update `expect-test` do be the same, that would be great.

MCJOHN974 replied:
Hmm, I don't fully understand how it should be

nagisa replied:
See the dependencies above such as `wasmparser.workspace = true`. For these dependencies the actual versions and such are maintained in a single place in the `Cargo.toml` file at the root of the project.

These should be the same.

MCJOHN974 replied:
Fixed [here](https://github.com/near/wasmtime/pull/40/commits/c13f5989ca1b6366e4106008333efe3baf1e0cc5)

nagisa commented on cranelift/codegen/src/isa/zkasm/inst/emit.rs:
```diff
@@ -466,10 +477,21 @@ impl MachInstEmit for Inst {
-                    sink,
-                );
+                // it is discussable if MulArith should be separate inst or part of ALURRR
+                // also, should I somehow show that we are using C register here?
```
Yeah, AluRRR is not at all appropriate in this instance. This first, and foremost, is something RISC-V specific. It has ALU instructions with three register arguments, which all can be arbitrary GPRs: two inputs and one output. Hence the RRR.

`:ARITH` instruction, on the other hand does not provide that flexibility in register allocation. It requires that the arguments are placed in the very specific registers before it executes, and these registers are *implicit* uses, and there are 3 of them, so what you’d really need is actually `AluRRRR` if we were going for abstracting away this pattern. This mismatch is also why you're finding yourself clobbering the C register in the implementation here – doing so is not valid as the register allocator might have assigned a value to C earlier and could be expecting this value to survive an execution of the `ARITH` instruction.

I would say that in general if you encounter artifacts from the RISC-V world, treat them with suspicion and default to removing and replacing them with something fresh.

nagisa reviewed (commented):
Hope the inline comment answers your questions? Let me know if what I wrote isn’t clear enough, but in summary, I believe that if you add a `Inst::MulArith` with 4 operands, you will naturally be forced to account for everything that might come up, including marking the `C` register as the use.

MCJOHN974 commented:
@nagisa hey, I separated ``MulArith`` from ``ALURRR`` and, I think, properly marked ``C`` register as clobbered as you can see in ``/cranelift/zkasm_data/generated/mul.zkasm``. But I don't understand how you want to use ``.workspace`` dependencies

MCJOHN974 commented:
> @nagisa hey, I separated `MulArith` from `ALURRR` and, I think, properly marked `C` register as clobbered as you can see in `/cranelift/zkasm_data/generated/mul.zkasm`. But I don't understand how you want to use `.workspace` dependencies

Or we can separate adding logging to another PR

nagisa commented on cranelift/codegen/src/isa/zkasm/inst/mod.rs:
```diff
@@ -269,6 +270,17 @@ fn zkasm_get_operands<F: Fn(VReg) -> VReg>(inst: &Inst, collector: &mut OperandC
+            collector.reg_fixed_use(rs1, a0());
+            collector.reg_fixed_use(rs2, b0());
+
+            // TODO: are next 3 lines are proper way to show that register C is clobbered???
```
Yes, this looks exactly right for the `0 => C` line in there.

nagisa commented on cranelift/codegen/src/isa/zkasm/inst/mod.rs:
```diff
@@ -955,7 +967,10 @@ impl Inst {
             }
-
+            &Inst::MulArith { rd, rs1, rs2 } => {
+                // TODO: I am not very understand what should be here
```
This is the disassembly of the instruction. Should output the textual representation of the instruction as well.

nagisa commented on cranelift/zkasm_data/mul.wat:
```diff
@@ -0,0 +1,9 @@
+ (import "env" "assert_eq" (func $assert_eq (param i32) (param i32)))
+ (func $main
+	i32.const 31
+	i32.const 64516129
```
Can you also add a test with `i64` multiplications? The 32-bit ones I believe should "just" work out of the box with plain `A * B`, but 64-bit ones AFAIU would require the `:ARITH`?

MCJOHN974 replied:
But for this numbers ``*`` still works

MCJOHN974 replied:
Oh, I found test which is impossible without arith: [commit](https://github.com/near/wasmtime/pull/40/commits/abb55c2be40e044d8083aabca318d79c3ebdd474). But also find a bug (=

MCJOHN974 commented:
@nagisa, I removed TODOs, and fixed dependencies according to your advice. Added test which will work only with using of ``ARITH``. Also, left a comment about work we should do when zkvm turns to 64 bit

nagisa commented on cranelift/zkasm_data/i64_mul.wat:
```diff
@@ -0,0 +1,15 @@
+    i64.const 214748366
+    i64.const 2
+	i64.mul
+    i64.mul
```
Hm, I worry that might be comparing two numbers that have simply wrapped around together to the same value. i.e. asserting that the multiplication result is `46116860270173224 % 2**32`, rather than `46116860270173224` in full.)

We better find how to represent `i64.const` proper in zkasm, sounds like a good question to ask polygon folks next week. In the mean time I think it is fine to merge this and perhaps file an issue for a TODO here?

MCJOHN974 replied:
Added this to the agenda of upcoming meeting. Also created an [issue](https://github.com/near/wasmtime/issues/53) to remind fix this after we get the answer
